### DISCUSSION TITLE
Additional 'createdAt' TTL index in MongoDB collections

### DISCUSSION THREAD

**Discussion by sarunasandr** [sorry-cypress/sorry-cypress#762] [Ideas]
### Problem
Director and API services work with MongoDB and store data about runs and other details. After running Sorry-Cypress for some time, I noticed that MongoDB storage usage keeps growing until it reaches storage limits. This happens because there is no TTL for the data and it is kept forever. Normally, test runs details only need to persist in a database for a finite amount of time.

I did some research and I found that MongoDB can have TTL indexes (https://www.mongodb.com/docs/manual/core/index-ttl/), but there is a requirement - [field value should be a date or an array that contains date values](https://www.mongodb.com/docs/manual/reference/bson-types/#std-label-document-bson-type-date). Currently, Sorry-Cypress uses collections which doesn't have any index with date inside, so MongoDB TTL indexes can't be created.

### Proposed solution
Add additional TTL index 'createdAt' for records in collections when creating DB entries and allow to specify [expireAfterSeconds](https://www.mongodb.com/docs/manual/core/index-ttl/#:~:text=and%20specify%20the-,expireAfterSeconds,-option%20with%20the) value. This would allow people to define how long they want to see test runs and avoid problems by reaching MongoDB storage limits.

**Comment by mathpaquette** [comment]
Hi,

`Normally, test runs details only need to persist in a database for a finite amount of time. `

What do you mean ?

**Reply by sarunasandr** [reply]
Hello,

I mean in our case, tests that were run months ago can be deleted and not kept in database. Currently, we don't have any control over how long they will be stored in the database.

**Comment by sarunasandr** [comment]
Hello,

Any news on this?

**Comment by redbugz** [comment]
We just ran into this issue as well, recording started to fail because we hit our Mongo limits. We don't want to pay for a huge Mongo instance, so some way to expire or batch delete test runs that are old would be nice. Or some way to set size limits and have old runs deleted as new ones are created, similar to a FIFO queue.

Looking at our data, the bulk of the size is in the instance collection, which stores the data for every single spec. By deleting some of those, the rest of the UI still works, it just shows no data when you click on an individual spec file in a run.
For some, they probably want to delete the entire run along with the related data.

I can see 2 possible approaches, the proposed one here that is just a TTL for runs (or just instances if people are okay with older runs losing the details). The other is perhaps to use [Capped Collections](https://www.mongodb.com/docs/v6.0/core/capped-collections/) to set a size limit, and it works like a circular buffer overwriting older data with newer data.

**Comment by agoldis** [comment]
There was a PR long time ago that was trying the TTL approach, but I required the field to be a proper `Date`, it's a `string` today (my bad)
